### ISSUE TITLE
Authentication with serviceaccounts doesn't work

### ISSUE THREAD

**Issue by Smana** [nats-io/nats-operator#129]
Hey guys,

I'm currently testing the nats operator and I would like to use the serviceaccounts to manage the permissions as stated in the documentation.
We're using the kubernetes version 1.13.2 so the token request API is already enabled.

I deployed the nats-operator using the helm chart, pretty straightforward. I just changed 2 values as follows:
```
image:
  tag: 0.4.0-v1alpha2
auth:
  enableServiceAccounts: true
```

Then I wanted to test the permissions using natsserviceroles and the example from the documentation:
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: nats-admin-user
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: nats-user
---
apiVersion: nats.io/v1alpha2
kind: NatsServiceRole
metadata:
  name: nats-user
  namespace: default

  # Specifies which NATS cluster will be mapping this account.
  labels:
    nats_cluster: nats-cluster
spec:
  permissions:
    publish: ["foo.*", "foo.bar.quux"]
    subscribe: ["foo.bar"]
---
apiVersion: nats.io/v1alpha2
kind: NatsServiceRole
metadata:
  name: nats-admin-user
  namespace: default
  labels:
    nats_cluster: nats-cluster
spec:
  permissions:
    publish: [">"]
    subscribe: [">"]
```
```
kubectl apply -f natservicerole.yaml 
serviceaccount/nats-admin-user created
serviceaccount/nats-user created
natsservicerole.nats.io/nats-user created
natsservicerole.nats.io/nats-admin-user created
```

And I created the test pod (using the newly created secret):
```
kubectl get secret | grep nats-user  
nats-user-token-44q4f                   kubernetes.io/service-account-token   3      114s
```
```
apiVersion: v1
kind: Pod
metadata:
  name: nats-user-pod
  labels:
    nats_cluster: nats-cluster
spec:
  volumes:
    - name: "token"
      projected:
        sources:
        - secret:
            name: "nats-user-token-44q4f"
            items:
              - key: token
                path: "token"
  restartPolicy: Never
  containers:
    - name: nats-ops
      command: ["/bin/sh"]
      image: "wallyqs/nats-ops:latest"
      tty: true
      stdin: true
      stdinOnce: true
      volumeMounts:
      - name: "token"
        mountPath: "/var/run/secrets/nats.io"
```
```
/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@nats-cluster:4222 foo.bar
Listening on [foo.bar]
```
However using another user works too ....
```
/go # nats-sub -s nats://nats-admin-user:`cat /var/run/secrets/nats.io/token`@nats-cluster:4222 foo.bar
Listening on [foo.bar]
```
I had a look to the nats.conf secret and nothing seems changed, we still have the same default config:
```
kubectl get secret nats-cluster -o yaml
echo "<b64>" | base64 -d
{
  "port": 4222,
  "http_port": 8222,
  "cluster": {
    "port": 6222,
    "routes": [
      "nats://nats-cluster-1.nats-cluster-mgmt.default.svc:6222",
      "nats://nats-cluster-2.nats-cluster-mgmt.default.svc:6222",
      "nats://nats-cluster-3.nats-cluster-mgmt.default.svc:6222"
    ]
  },
  "authorization": {}
}
```

**Comment by wallyqs** [comment]
🤔 ok couple of things...

- When doing `kubectl get secrets`, do you see a list of bound token objects? There should be a few created, for example:

```
 kubectl get secrets | grep bound
nats-admin-user-nats-cluster-bound-token   Opaque                                1         1m
nats-user-nats-cluster-bound-token         Opaque                                1         1m
```

- In the helm chart `configReload.enabled` should be enabled, or `enableConfigReload: true` if not using the helm chart full example below:

```sh
echo '
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: nats-admin-user
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: nats-user
---
apiVersion: nats.io/v1alpha2
kind: NatsServiceRole
metadata:
  name: nats-user
  namespace: default

  # Specifies which NATS cluster will be mapping this account.
  labels:
    nats_cluster: nats-cluster
spec:
  permissions:
    publish: ["foo.*", "foo.bar.quux"]
    subscribe: ["foo.bar"]
---
apiVersion: nats.io/v1alpha2
kind: NatsServiceRole
metadata:
  name: nats-admin-user
  namespace: default
  labels:
    nats_cluster: nats-cluster
spec:
  permissions:
    publish: [">"]
    subscribe: [">"]
---
apiVersion: "nats.io/v1alpha2"
kind: "NatsCluster"
metadata:
  name: "nats-cluster"
spec:
  size: 3
  version: "1.4.0"
  pod:
    enableConfigReload: true
    reloaderImage: connecteverything/nats-server-config-reloader
  auth:
    enableServiceAccounts: true
' | kubectl apply -f -
```

**Comment by Smana** [comment]
Thx @wallyqs , I deployed again the operator with configReload enabled and I'm now seing configuration changes : 
```
{
  "port": 4222,
  "http_port": 8222,
  "cluster": {
    "port": 6222,
    "routes": [
      "nats://nats-cluster-1.nats-cluster-mgmt.default.svc:6222",
      "nats://nats-cluster-3.nats-cluster-mgmt.default.svc:6222"
    ]
  },
  "authorization": {
    "users": [
      {
        "username": "nats-admin-user",
        "permissions": {
          "publish": [
            ">"
          ],
          "subscribe": [
            ">"
          ]
        }
      },
      {
        "username": "nats-user",
        "permissions": {
          "publish": [
            "foo.*",
            "foo.bar.quux"
          ],
          "subscribe": [
            "foo.bar"
          ]
        }
      }
    ]
  }
}
```

However I still don't have `bound` secrets and I'm still able to listen with both users
```
/go # nats-sub -s nats://nats-user:`cat /var/run/secrets/nats.io/token`@nats-cluster:4222 foo.bar
Listening on [foo.bar]
^C
/go # nats-sub -s nats://nats-admin-user:`cat /var/run/secrets/nats.io/token`@nats-cluster:4222 foo.bar
Listening on [foo.bar]
^C
```

**Comment by Smana** [comment]
I just saw that one of the nats replicas doesn't start and i've the following error in the reloader logs : ```2019/02/23 10:09:46 Starting NATS Server Reloader v0.2.2-v1alpha2+git
2019/02/23 10:09:46 Error: open /var/run/nats/gnatsd.pid: no such file or directory
2019/02/23 10:09:48 Error: open /var/run/nats/gnatsd.pid: no such file or directory
2019/02/23 10:09:50 Error: open /var/run/nats/gnatsd.pid: no such file or directory
2019/02/23 10:09:52 Error: open /var/run/nats/gnatsd.pid: no such file or directory
2019/02/23 10:09:54 Error: open /var/run/nats/gnatsd.pid: no such file or directory
2019/02/23 10:09:56 Error: open /var/run/nats/gnatsd.pid: no such file or directory
Error: Too many errors attempting to find server process
```

**Comment by Smana** [comment]
Another error log in the operator pod : 
```
E0223 09:55:51.184438       1 generic.go:108] error syncing "default/nats-cluster": failed to reconcile pods: pod "nats-cluster-2" has been deleted                                                               
time="2019-02-23T10:09:39Z" level=info msg="waiting for pod \"nats-cluster-2\" to become ready" cluster-name=nats-cluster pkg=cluster                                                                             
E0223 10:14:39.667057       1 generic.go:108] error syncing "default/nats-cluster": failed to reconcile pods: timed out waiting for the condition 
```

**Comment by Smana** [comment]
I tried to delete one of the replicas but it's not coming back. It doesn't seem to be controlled by any replicaset/statefulset ...

**Comment by wallyqs** [comment]
The operator acts as the controller instead of statefulset/replicaset.  I could reproduce this issue on a v1.12 cluster, it occurs when the operator fails to make a token request (getting a resource not found error token for example) so it creates a user without a password and then server fails to start due to pass for a user being required.

**Comment by Smana** [comment]
Hello, after some changes with @wallyqs help, here are some details about my issue.
First of all some options were missing in my apiserver config:
```
    - --service-account-signing-key-file=/etc/ssl/private/kube-apiserver-server-key.pem
    - --service-account-issuer=api
    - --api-audiences=api
```
After this change the config file (secret) that is used by the nats nodes was updated.
```
        "username": "nats-user",
        "password": "eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJhdWQiOlsibmF0czovL25hdHMtY2x1c3Rlci5kZWZhdWx0LnN2YyJdLCJleHAiOjE1NTIxMjA0NzIsImlhdCI6MTU1MjExNjg3MiwiaXNzIjoiYXBpIiwia3ViZXJuZXRlcy5pbyI6eyJuYW1lc3BhY2UiOiJkZWZhdWx0Iiwic2VjcmV0Ijp7Im5hbWUiOiJuYXRzLXVzZXItbmF0cy1jbHVzdGVyLWJvdW5kLXRva2VuIiwidWlkIjoiYzdlNDNkYTYtNDIzZC0xMWU5LThiMzctNTQ5ZjM1MWFjNGYwIn0sInNlcnZpY2VhY2NvdW50Ijp7Im5hbWUiOiJuYXRzLXVzZXIiLCJ1aWQiOiIwMjEzZWQ5Yi00MWJhLTExZTktOGIzNy01NDlmMzUxYWM0ZjAifX0sIm5iZiI6MTU1MjExNjg3Miwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmRlZmF1bHQ6bmF0cy11c2VyIn0.SGdLN5PFS78USkRyGM64ZieVwDrB0xWUzL99Shb0AExnLG3Yg8Fi2NuDBUBWJlzoWNgnP3wYQuyqI1HxEQCtSALcb8wWsay8NAZxq1jNL10UxXy67_I__295kMxe4GwsVRpCLojM1BDptqyen2-oR5XylPvRyTCNBMJuBrXfB3zsjhh99WVthAXJukz7OGJGy2Mm6wMiHMRQscLWQJOsaNNSyEdFUGRjW2EFBAvChPn2f2u-z3Q1TIJxbdLrrHuxuTwEvx8k_XH_yTgjWhuHtd4TXq8C3EA0FmJDXycFSaWN7fyMQn5V5r57ScLwV79VSSc1L6nCKw1bC3cNwQIYsw",
        "permissions": {
          "publish": [
            "foo.*",
            "foo.bar.quux"
          ],
          "subscribe": [
            "foo.bar"
          ]
        }
      }
```

However there are no secrets with a names that contains `bound`
```
kubectl get secrets
NAME                          TYPE                                  DATA   AGE
default-token-rqplb           kubernetes.io/service-account-token   3      50d
nats-admin-user-token-wxlwj   kubernetes.io/service-account-token   3      15h
nats-clients-tls              Opaque                                3      13h
nats-cluster                  Opaque                                1      13h
nats-operator-token-wxp8l     kubernetes.io/service-account-token   3      15h
nats-routes-tls               Opaque                                3      13h
nats-server-token-6bcp4       kubernetes.io/service-account-token   3      15h
nats-user-token-h2rwk         kubernetes.io/service-account-token   3      15h
```
Looking at the apiserver audit logs we can see that these secrets are deleted immediately after being created for an unknown reason.
Here are the apiserver logs for the secret `nats-user-nats-cluster-bound-token` (note: I just deleted a few get commands, too many of them)
https://gist.github.com/Smana/03020a19ab055238aaa907a25457ce8c

Tell me if you need further details.
As a reminder:
kubernetes v1.13.2 on-premises

**Comment by Smana** [comment]
I think I found out how to workaround the issue:
It seems that you cannot define both user/password and serviceaccount authentication.
I just commented out the following helm values:
```
  ## This is where you enter a username/password for 1 user
  # username: "my-user"
  # password: "T0pS3cr3t"
```
And now I can use my bound tokens ! :muscle: 
```
kubectl get secrets -n nats-io                                                                                  
NAME                                       TYPE                                  DATA   AGE
default-token-7j7hf                        kubernetes.io/service-account-token   3      9h
nats-admin-user-nats-cluster-bound-token   Opaque                                1      4m51s
nats-admin-user-token-kmcds                kubernetes.io/service-account-token   3      9h
nats-clients-tls                           Opaque                                3      9h
nats-cluster                               Opaque                                1      5m30s
nats-io-nats-operator-clients-auth         Opaque                                1      5m50s
nats-operator-token-tsmth                  kubernetes.io/service-account-token   3      5m50s
nats-routes-tls                            Opaque                                3      9h
nats-user-nats-cluster-bound-token         Opaque                                1      4m53s
nats-user-token-rqzqw                      kubernetes.io/service-account-token   3      9h
```
I don't know if that's an expected behaviour @wallyqs

**Comment by Smana** [comment]
I just checked this morning and they were deleted this morning. I don't know why.

**Comment by wallyqs** [comment]
Thanks @Smana for continuing to update the issue, I think I have found the bug that was causing this: https://github.com/nats-io/nats-operator/pull/144

**Comment by Smana** [comment]
Hi @wallyqs I've tested your change and that works great. Thanks !
I guess we can close this issue when the change will be merged

**Comment by wallyqs** [comment]
Fixed via #144
